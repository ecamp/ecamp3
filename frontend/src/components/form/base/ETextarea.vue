<template>
  <ValidationProvider
    v-slot="{ errors: veeErrors }"
    tag="div"
    :name="name"
    :vid="veeId"
    :rules="veeRules"
    :required="required"
    class="e-form-container"
  >
    <v-tiptap-editor
      v-bind="$attrs"
      :with-extensions="false"
      :filled="filled"
      :hide-details="hideDetails"
      :error-messages="veeErrors.concat(errorMessages)"
      :label="label || name"
      :class="[inputClass]"
      v-on="$listeners"
    >
      <!-- passing through all slots -->
      <slot v-for="(_, name) in $slots" :slot="name" :name="name" />
      <template v-for="(_, name) in $scopedSlots" :slot="name" slot-scope="slotData">
        <slot :name="name" v-bind="slotData" />
      </template>
    </v-tiptap-editor>
  </ValidationProvider>
</template>

<script>
import { ValidationProvider } from 'vee-validate'
import { formComponentPropsMixin } from '@/mixins/formComponentPropsMixin.js'
import VTiptapEditor from '@/components/form/tiptap/VTiptapEditor.vue'
import { formComponentMixin } from '@/mixins/formComponentMixin.js'

export default {
  name: 'ETextarea',
  components: { VTiptapEditor, ValidationProvider },
  mixins: [formComponentPropsMixin, formComponentMixin],
}
</script>
