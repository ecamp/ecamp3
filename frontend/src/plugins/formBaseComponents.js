// Global registration of all Vue components in folder components/form/base

import ECheckbox from '@/components/form/base/ECheckbox.vue'
import EColorPicker from '@/components/form/base/EColorPicker.vue'
import EDatePicker from '@/components/form/base/EDatePicker.vue'
import EForm from '@/components/form/base/EForm.vue'
import ERichtext from '@/components/form/base/ERichtext.vue'
import ESelect from '@/components/form/base/ESelect.vue'
import ESwitch from '@/components/form/base/ESwitch.vue'
import ETextarea from '@/components/form/base/ETextarea.vue'
import ETextField from '@/components/form/base/ETextField.vue'
import ETimePicker from '@/components/form/base/ETimePicker.vue'

class FormBaseComponentsPlugin {
  install(Vue) {
    Vue.component('ECheckbox', ECheckbox)
    Vue.component('EColorPicker', EColorPicker)
    Vue.component('EDatePicker', EDatePicker)
    Vue.component('EForm', EForm)
    Vue.component('ERichtext', ERichtext)
    Vue.component('ESelect', ESelect)
    Vue.component('ESwitch', ESwitch)
    Vue.component('ETextarea', ETextarea)
    Vue.component('ETextField', ETextField)
    Vue.component('ETimePicker', ETimePicker)
  }
}

export default new FormBaseComponentsPlugin()
