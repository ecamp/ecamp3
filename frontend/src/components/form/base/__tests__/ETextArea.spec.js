import Vue from 'vue'
import Vuetify from 'vuetify'

import formBaseComponents from '@/plugins/formBaseComponents'

import { mount as mountComponent } from '@vue/test-utils'
import ETextarea from '../ETextarea.vue'
import { mockEventClass } from '@/test/mockEventClass'

Vue.use(Vuetify)
Vue.use(formBaseComponents)

mockEventClass('ClipboardEvent')
mockEventClass('DragEvent')

describe('An ETextArea', () => {
  let vuetify

  const multiLineText = `
    Here comes a text
    with new lines
    and new lines with \n in them
    and tags <i>a</i>
    `

  const mount = (
    options,
    template = `
        <div data-app>
          <e-textarea v-model="data"/>
        </div>
      `
  ) => {
    const app = Vue.component('App', {
      components: { ETextarea },
      data: () => ({ data: null }),
      template: template,
    })
    return mountComponent(app, { vuetify, attachTo: document.body, ...options })
  }
  beforeEach(() => {
    vuetify = new Vuetify()
  })
  test('looks like a textarea', async () => {
    const wrapper = mount()
    expect(wrapper).toMatchSnapshot('notext')

    await wrapper.setData({ data: multiLineText })
    expect(wrapper).toMatchSnapshot('withtext')

    const mountWithMultiLine = mount(
      { data: () => ({ data: multiLineText }) },
      `
        <div data-app>
          <e-textarea v-model="data" multi-line/>
        </div>
      `
    )
    expect(mountWithMultiLine).toMatchSnapshot('multiline')

    const mountAsinControls = mount(
      { data: () => ({ data: multiLineText }) },
      `
        <div data-app>
          <e-textarea
                v-model="data"
                :rows="3"
                auto-grow
          />
        </div>
      `
    )
    expect(mountAsinControls).toMatchSnapshot('mountAsInControls')
  })

  test('updates the text with the viewmodel', async () => {
    const wrapper = mount()
    await wrapper.setData({ data: multiLineText })
    const textWithoutMultiLine = multiLineText
      .replace(/\n\s*/g, ' ')
      .replace('  ', ' ')
      .replace('<i>', '')
      .replace('</i>', '')
      .trim()
    expect(wrapper.find('.editor__content').text()).toBe(textWithoutMultiLine)
    expect(wrapper.find('.e-form-container').element.getAttribute('value')).toBe(
      multiLineText
    )
    expect(wrapper.vm.data).toBe(multiLineText)
  })
})
