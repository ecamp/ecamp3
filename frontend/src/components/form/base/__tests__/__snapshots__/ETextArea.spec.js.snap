// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`An ETextArea > looks like a textarea > mountAsInControls 1`] = `
<div data-app>
     <div auto-grow
          class="e-form-container"
          rows="3"
          value="
    Here comes a text
    with new lines
    and new lines with 
 in them
    and tags <i>a</i>
    ">
          <div class="v-input v-input--hide-details v-input--is-label-active v-input--is-dirty theme--light v-text-field v-text-field--single-line v-text-field--filled v-text-field--enclosed">
               <div class="v-input__control">
                    <div class="v-input__slot">
                         <div class="v-text-field__slot">
                              <div auto-grow
                                   class="editor"
                                   id
                                   rows="3">
                                   <!---->
                                   <div class="editor__content"></div>
                              </div>
                         </div>
                    </div>
               </div>
          </div>
     </div>
</div>
`;

exports[`An ETextArea > looks like a textarea > multiline 1`] = `
<div data-app>
     <div class="e-form-container"
          multi-line
          value="
    Here comes a text
    with new lines
    and new lines with 
 in them
    and tags <i>a</i>
    ">
          <div class="v-input v-input--hide-details v-input--is-label-active v-input--is-dirty theme--light v-text-field v-text-field--single-line v-text-field--filled v-text-field--enclosed">
               <div class="v-input__control">
                    <div class="v-input__slot">
                         <div class="v-text-field__slot">
                              <div class="editor"
                                   id
                                   multi-line>
                                   <!---->
                                   <div class="editor__content"></div>
                              </div>
                         </div>
                    </div>
               </div>
          </div>
     </div>
</div>
`;

exports[`An ETextArea > looks like a textarea > notext 1`] = `
<div data-app>
     <div class="e-form-container">
          <div class="v-input v-input--hide-details theme--light v-text-field v-text-field--single-line v-text-field--filled v-text-field--enclosed">
               <div class="v-input__control">
                    <div class="v-input__slot">
                         <div class="v-text-field__slot">
                              <div class="editor"
                                   id>
                                   <!---->
                                   <div class="editor__content"></div>
                              </div>
                         </div>
                    </div>
               </div>
          </div>
     </div>
</div>
`;

exports[`An ETextArea > looks like a textarea > withtext 1`] = `
<div data-app>
     <div class="e-form-container"
          value="
    Here comes a text
    with new lines
    and new lines with 
 in them
    and tags <i>a</i>
    ">
          <div class="v-input v-input--hide-details v-input--is-label-active v-input--is-dirty theme--light v-text-field v-text-field--single-line v-text-field--filled v-text-field--enclosed">
               <div class="v-input__control">
                    <div class="v-input__slot">
                         <div class="v-text-field__slot">
                              <div class="editor"
                                   id>
                                   <!---->
                                   <div class="editor__content">
                                        <div class="tiptap ProseMirror"
                                             contenteditable="true"
                                             tabindex="0"
                                             translate="no">
                                             <p>Here comes a text with new lines and new lines with in them and tags a</p>
                                        </div>
                                   </div>
                              </div>
                         </div>
                    </div>
               </div>
          </div>
     </div>
</div>
`;
